sw/device/lib/dif/dif_aes.c

```diff
@@ -0,0 +1,424 @@
+    dif_aes_key_share_t key, dif_aes_iv_t iv) {
+  if (aes == NULL || transaction == NULL) {
+    return kDifAesStartBadArg;
+  }
+
+  if (!aes_idle(aes)) {
+    return kDifAesStartBusy;
+  }
+
+  dif_aes_start_result_t result =
+      configure(aes, transaction, kAesModeFieldValCtr);
+  if (result != kDifAesStartOk) {
+    return result;
+  }
+
+  aes_set_multireg(aes, &key.part0[0], AES_KEY_SHARE0_MULTIREG_COUNT,
+                   AES_KEY_SHARE0_0_REG_OFFSET);
+  aes_set_multireg(aes, &key.part1[0], AES_KEY_SHARE1_MULTIREG_COUNT,
+                   AES_KEY_SHARE1_0_REG_OFFSET);
+  aes_set_multireg(aes, &iv.iv[0], AES_IV_MULTIREG_COUNT, AES_IV_0_REG_OFFSET);
+
+  return kDifAesStartOk;
+}
+
+dif_aes_end_result_t dif_aes_end(const dif_aes_t *aes) {
+  if (aes == NULL) {
+    return kDifAesEndBadArg;
+  }
+
+  if (!aes_idle(aes)) {
+    return kDifAesEndBusy;
+  }
+
+  aes_clear_internal_state(aes);
+
+  return kDifAesEndOk;
+}
+
+dif_aes_load_data_result_t dif_aes_load_data(const dif_aes_t *aes,
+                                             const dif_aes_data_t data) {
+  if (aes == NULL) {
+    return kDifAesLoadDataBadArg;
+  }
+
+  if (!aes_input_ready(aes)) {
+    return kDifAesLoadDataBusy;
+  }
+
+  aes_set_multireg(aes, &data.data[0], AES_DATA_IN_MULTIREG_COUNT,
+                   AES_DATA_IN_0_REG_OFFSET);
+
+  return kDifAesLoadDataOk;
+}
+
+dif_aes_read_output_result_t dif_aes_read_output(const dif_aes_t *aes,
+                                                 dif_aes_data_t *data) {
+  if (aes == NULL || data == NULL) {
+    return kDifAesReadOutputBadArg;
+  }
+
+  if (!aes_output_valid(aes)) {
+    return kDifAesReadOutputInvalid;
+  }
+
+  for (int i = 0; i < AES_DATA_OUT_MULTIREG_COUNT; ++i) {
+    ptrdiff_t offset = AES_DATA_OUT_0_REG_OFFSET + (i * sizeof(uint32_t));
+
+    data->data[i] = mmio_region_read32(aes->params.base_addr, offset);
+  }
+
+  return kDifAesReadOutputOk;
+}
+
+dif_aes_result_t dif_aes_trigger(const dif_aes_t *aes,
+                                 dif_aes_trigger_t trigger) {
+  if (aes == NULL) {
+    return kDifAesBadArg;
+  }
+
+  uint32_t reg;
+  switch (trigger) {
+    case kDifAesTriggerStart:
+      reg = bitfield_bit32_write(0, AES_TRIGGER_START_BIT, true);
+      break;
+    case kDifAesTriggerKeyIvDataInClear:
+      reg = bitfield_bit32_write(0, AES_TRIGGER_KEY_IV_DATA_IN_CLEAR_BIT, true);
+      break;
+    case kDifAesTriggerDataOutClear:
+      reg = bitfield_bit32_write(0, AES_TRIGGER_DATA_OUT_CLEAR_BIT, true);
+      break;
+    case kDifAesTriggerPrngReseed:
+      reg = bitfield_bit32_write(0, AES_TRIGGER_PRNG_RESEED_BIT, true);
+      break;
+    default:
+      return kDifAesError;
+  }
+
+  mmio_region_write32(aes->params.base_addr, AES_TRIGGER_REG_OFFSET, reg);
+
+  return kDifAesOk;
+}
+
+dif_aes_result_t dif_aes_status_get(const dif_aes_t *aes, dif_aes_status_t flag,
+                                    bool *set) {
```

Comment:
For a followup, it seems like this wants to return a bitset...?

Reply:
Possibly, I will leave it with @alphan to decide.